Prevent not-intended loading of `ActionDispatch::IntegrationTest`

yui-knk opened the pull request:
After 9d378747326d26cf1afdac4433ead22967af0984 `ActionDispatch::IntegrationTest`
class is loaded and defined in all Rails environments, not only test but also
production. This is not-intended loading of a class which is only used in
test environment.

kaspth commented:
Think the delivery reset is a good idea, so don't want to back track on it. Shouldn't we just scope the resets to the test environment or similar? cc @senny.

rafaelfranca commented:
We could even define a `on_load` hook so the module is only included when the class is loaded.

yui-knk commented:
@rafaelfranca Do you mean defining `action_dispatch_integration_test` hook? If so , I agree!

I will update to

* define `ActiveSupport.run_load_hooks(:action_dispatch_integration_test, self)` in actionpack/lib/action_dispatch/testing/integration.rb
* use `ActiveSupport.on_load(:action_dispatch_integration_test)` in actionmailer/lib/action_mailer/railtie.rb

rafaelfranca commented:
:+1: exactly this.

yui-knk committed:
Prevent not-intended loading of `ActionDispatch::IntegrationTest`

After 9d378747326d26cf1afdac4433ead22967af0984 `ActionDispatch::IntegrationTest`
class is loaded and defined in all Rails environments, not only test but also
production. This is not-intended loading of a class which is only used in
test environment.
To prevent not-intended loading, add `ActiveSupport.run_load_hooks` to
`ActionDispatch::IntegrationTest` with `action_dispatch_integration_test` name
and use it in `ActionMailer`.

yui-knk commented:
Updated!! :heart_eyes:

senny commented on actionmailer/lib/action_mailer/railtie.rb:
```diff
@@ -45,9 +45,9 @@ class Railtie < Rails::Railtie # :nodoc:
-        ActionDispatch::IntegrationTest.send :include, ActionMailer::TestCase::ClearTestDeliveries
       end
+
+      ActiveSupport.on_load(:action_dispatch_integration_test) { self.include ActionMailer::TestCase::ClearTestDeliveries }
```
can we remove `self.`?

yui-knk replied:
sure!

senny commented:
The hook seems to be a good solution. I can imagine that other gems might want to use that too.

senny commented:
@yui-knk can you add a changelog entry as well? Not so much for the bug fix but for the added hook.

yui-knk commented:
OK, I wil  add changelog for the added hook.

yui-knk commented:
Add changelog :)

senny commented:
@yui-knk thank you :yellow_heart:

kaspth commented:
Well done ❤️

yui-knk commented:
Thanks for your review :heart_eyes:

amatsuda commented:
Nice! :+1:
